etcd-runner: fails on etcd-tester (election, no progress after 1 minute)

```
2017-05-05 20:38:04.251694 I | etcd-tester: [round#0 case#2] recovered failure
2017-05-05 20:38:04.251726 I | etcd-tester: [round#0 case#2] pausing the stressers...
2017-05-05 20:38:04.252160 I | etcd-tester: lease stresser "10.240.0.2:2379" is closed
2017-05-05 20:38:04.252322 I | etcd-tester: keyStresser "10.240.0.2:2379" is closed
2017-05-05 20:38:04.254374 I | etcd-tester: lease stresser "10.240.0.4:2379" is closed
2017-05-05 20:38:04.254689 I | etcd-tester: lease stresser "10.240.0.6:2379" is closed
2017-05-05 20:38:04.254839 I | etcd-tester: lease stresser "10.240.0.3:2379" is closed
2017-05-05 20:38:04.255294 I | etcd-tester: lease stresser "10.240.0.5:2379" is closed
2017-05-05 20:38:04.255727 I | etcd-tester: keyStresser "10.240.0.4:2379" is closed
2017-05-05 20:38:04.255819 I | etcd-tester: keyStresser "10.240.0.6:2379" is closed
2017-05-05 20:38:04.255859 I | etcd-tester: keyStresser "10.240.0.5:2379" is closed
2017-05-05 20:38:04.255906 I | etcd-tester: keyStresser "10.240.0.3:2379" is closed
2017-05-05 20:38:04.255930 I | etcd-tester: [round#0 case#2] paused stressers
2017-05-05 20:38:04.255938 I | etcd-tester: [round#0 case#2] wait until cluster is healthy
2017-05-05 20:38:04.283546 I | etcd-tester: [round#0 case#2] cluster is healthy
2017-05-05 20:38:04.283572 I | etcd-tester: [round#0 case#2] checking consistency and invariant of cluster
2017-05-05 20:38:10.325973 I | etcd-tester: [round#0 case#2] (/home/gyuho/go/bin/etcd-runner [election 1494016619491935753 --dial-timeout=10s --endpoints 10.240.0.4:2379 --total-client-connections=10 --rounds=0 --req-rate 100]) stderr 20:37:59.977209 no progress after 1 minute!
panic: no progress after 1 minute!

goroutine 1 [running]:
log.Panic(0xc420339ad8, 0x1, 0x1)
	/usr/local/go/src/log/log.go:322 +0xc0
```


> `etcd-tester --enable-pprof --etcd-runner=$etcd-runner --stresser=keys,lease,election-runner,watch-runner,lock-racer-runner,lease-runner --agent-endpoints="$(echo $AGENT_RPC_ENDPOINTS)"`

--- a/tools/functional-tester/etcd-runner/command/election_command.go
+++ b/tools/functional-tester/etcd-runner/command/election_command.go
@@ -89,14 +89,14 @@ func runElectionFunc(cmd *cobra.Command, args []string) {
 				}
 			}()
 			err = e.Campaign(ctx, v)
+			cancel()
+			<-donec
 			if err == nil {
 				observedLeader = v
 			}
 			if observedLeader == v {
 				validateWaiters = len(rcs)
 			}
-			cancel()
-			<-donec
 			select {
 			case <-ctx.Done():
 				return nil

--- a/tools/functional-tester/etcd-runner/command/election_command.go
+++ b/tools/functional-tester/etcd-runner/command/election_command.go
@@ -129,8 +129,10 @@ func runElectionFunc(cmd *cobra.Command, args []string) {
 				return err
 			}
 			if observedLeader == v {
-				close(nextc)
+				oldNextc := nextc
 				nextc = make(chan struct{})
+				close(oldNextc)
+
 			}
 			<-rcNextc
 			observedLeader = ""

--- a/tools/functional-tester/etcd-runner/command/global.go
+++ b/tools/functional-tester/etcd-runner/command/global.go
@@ -56,7 +56,6 @@ func newClient(eps []string, timeout time.Duration) *clientv3.Client {
 }
 
 func doRounds(rcs []roundClient, rounds int, requests int) {
-	var mu sync.Mutex
 	var wg sync.WaitGroup
 
 	wg.Add(len(rcs))
@@ -73,22 +72,16 @@ func doRounds(rcs []roundClient, rounds int, requests int) {
 				for rc.acquire() != nil { /* spin */
 				}
 
-				mu.Lock()
 				if err := rc.validate(); err != nil {
 					log.Fatal(err)
 				}
-				mu.Unlock()
 
 				time.Sleep(10 * time.Millisecond)
 				rc.progress++
 				finished <- struct{}{}
 
-				mu.Lock()
 				for rc.release() != nil { /* spin */
-					mu.Unlock()
-					mu.Lock()
 				}
-				mu.Unlock()
 			}
 		}(&rcs[i])
 	}

--- a/tools/functional-tester/etcd-runner/command/lock_racer_command.go
+++ b/tools/functional-tester/etcd-runner/command/lock_racer_command.go
@@ -18,6 +18,7 @@ import (
 	"context"
 	"errors"
 	"fmt"
+	"sync"
 
 	"github.com/coreos/etcd/clientv3/concurrency"
 
@@ -47,6 +48,8 @@ func runRacerFunc(cmd *cobra.Command, args []string) {
 
 	rcs := make([]roundClient, totalClientConnections)
 	ctx := context.Background()
+	// mu ensures validate and release funcs are atomic.
+	var mu sync.Mutex
 	cnt := 0
 
 	eps := endpointsFromFlag(cmd)
@@ -69,12 +72,16 @@ func runRacerFunc(cmd *cobra.Command, args []string) {
 		m := concurrency.NewMutex(s, racers)
 		rcs[i].acquire = func() error { return m.Lock(ctx) }
 		rcs[i].validate = func() error {
+			mu.Lock()
+			defer mu.Unlock()
 			if cnt++; cnt != 1 {
 				return fmt.Errorf("bad lock; count: %d", cnt)
 			}
 			return nil
 		}
 		rcs[i].release = func() error {
+			mu.Lock()
+			defer mu.Unlock()
 			if err := m.Unlock(ctx); err != nil {
 				return err
 			}

